crates/accelerate/src/sabre_swap/layer.rs

```diff
@@ -145,90 +165,110 @@ impl FrontLayer {
-    pub fn insert(&mut self, index: NodeIndex, qubits: &[VirtualQubit; 2]) -> bool {
-        let [a, b] = *qubits;
-        if self.nodes.insert(index, *qubits).is_none() {
-            self.qubits[a.index()].push(b);
-            self.qubits[b.index()].push(a);
-            true
-        } else {
-            false
-        }
+    pub fn push(&mut self, qubits: [PhysicalQubit; 2]) {
+        let [a, b] = qubits;
+        self.qubits[a.index()].push(b);
+        self.qubits[b.index()].push(a);
+        self.len += 1;
     }
 
     /// Calculate the score of applying the given swap, relative to not applying it.
-    pub fn score(&self, swap: [VirtualQubit; 2], layout: &NLayout, dist: &ArrayView2<f64>) -> f64 {
-        if self.nodes.is_empty() {
+    pub fn score(&self, swap: [PhysicalQubit; 2], dist: &ArrayView2<f64>) -> f64 {
+        if self.len == 0 {
             return 0.0;
         }
         let [a, b] = swap;
-        let p_a = a.to_phys(layout);
-        let p_b = b.to_phys(layout);
         let mut total = 0.0;
         for other in self.qubits[a.index()].iter() {
             // If the other qubit is also active then the score won't have changed, but since the
             // distance is absolute, we'd double count rather than ignore if we didn't skip it.
             if *other == b {
                 continue;
             }
-            let p_other = other.to_phys(layout);
-            total += dist[[p_b.index(), p_other.index()]] - dist[[p_a.index(), p_other.index()]];
+            total += dist[[b.index(), other.index()]] - dist[[a.index(), other.index()]];
         }
         for other in self.qubits[b.index()].iter() {
             if *other == a {
                 continue;
             }
-            let p_other = other.to_phys(layout);
-            total += dist[[p_a.index(), p_other.index()]] - dist[[p_b.index(), p_other.index()]];
+            total += dist[[a.index(), other.index()]] - dist[[b.index(), other.index()]];
         }
-        total / self.nodes.len() as f64
+        total / self.len as f64
     }
 
     /// Calculate the total absolute score of this set of nodes over the given layout.
-    pub fn total_score(&self, layout: &NLayout, dist: &ArrayView2<f64>) -> f64 {
-        if self.nodes.is_empty() {
+    pub fn total_score(&self, dist: &ArrayView2<f64>) -> f64 {
+        if self.len == 0 {
             return 0.0;
         }
-        self.nodes
-            .values()
-            .map(|&[a, b]| dist[[a.to_phys(layout).index(), b.to_phys(layout).index()]])
+        self.qubits
+            .iter()
+            .enumerate()
+            .map(move |(a_index, others)| {
+                others
+                    .iter()
+                    .map(|b| {
+                        let b_index = b.index();
+                        if a_index <= b_index {
+                            dist[[a_index, b_index]]
+                        } else {
+                            0.0
+                        }
```

Comment:
Is there an advantage to doing this vs a `filter_map`?

Reply:
I didn't think about it - the original code used `map` because it had an iterator that directly ran over nodes, so didn't need to worry about the double counting.  I'll swap it to `filter_map` if you prefer, and another possibility is making the outer `map` a `flat_map` and removing the internal `sum`?

Reply:
I guess it was more an efficiency question, I didn't know if adding a bunch of 0.0s was going to be faster than using `filter_map` or something. But, yeah using a `filter_map`, `flat_map`, and a single `sum` seems the more natural way to do this with iterators.

Reply:
Naively, I don't think you can't represent `Option<f64>` without a discriminant (assuming you care about propagating the bit pattern payload of qNaNs?) but given LLVM will have the full context of the ensuing operations, it can probably unpick everything into sensible code.  I had a go in https://godbolt.org/z/Y531qhn9r, but I think at that point it'll be branch prediction and memory-access patterns that dictate the speed more than anything else.  I'll try locally and just make it a `flat_map+filter_map` assuming it's not visible.

If this code turns out to be something we need to micro-optimise, the next thing I'd try would be enforcing the `Vec<Vec<others>>` stuff to always allocate the other qubits in groups of 4 or 8, and then get the compiler to emit always-vectorised code output (using the self-index in the outer vec to represent "missing" slots in the inner vec, to abuse `dist[[a, a]] == 0.0`).

Reply:
It's possible that a further re-organisation of the data structure to store more in order to have a direct iterator over each gate exactly once without branching might help as well, but I want to wait on that til we've looked at what we're going to do about building up the extended set as a delta, because this PR still offers a speed-up over the status quo, and I don't want to get into premature optimisation that might become obsolete as soon as we add the extra requirement that it must be possible to remove single gates from the `ExtendedSet`.